Storage runtime permissions resolves #644

jd-alexander opened the pull request:
Closes #644 

This pull request addresses the request for permissions at runtime needed to acesss the device's storage.

![odl1](https://user-images.githubusercontent.com/1509205/38647149-df5e3c5a-3db8-11e8-8eed-9607fe876b43.png)       ![odk2](https://user-images.githubusercontent.com/1509205/38647148-df448c2e-3db8-11e8-8202-5ce32fd08a6e.png)


#### What has been done to verify that this works as intended?
The main targets of this refactor were classes that were calling `createODKDirs` of the application class as a means of creating the root directories for Collect's main functionalities. These classes are the means by which the app is launched whether it be from an intent or the app drawer so I utilized @grzesiek2010 's super cool [Collect Tester ](https://github.com/grzesiek2010/collectTester) app :sunglasses: . I  verified that when activities are being launched externally and the app doesn't have the storage permissions granted, a prompt is presented. The two content providers were also modified and tested because the default before was for them to always be available for querying. With this change, they return default values until the user open's the app and grants the permissions. From an automated testing standpoint, these changes caused tests running on devices with API 23 and higher to fail so I introduced Espresso's [GrantPermissionRule](https://developer.android.com/reference/android/support/test/rule/GrantPermissionRule.html) to grant the test case the storage permissions before the tests are run. 

#### Why is this the best possible solution? Were any other approaches considered?
To deal with the permissions boilerplate I utilized the [Dexter permissions library](https://github.com/Karumi/Dexter). I could have utilized the [existing lib](https://github.com/googlesamples/easypermissions) but Dexter allows all the permissions related callbacks, methods and checks to be fully decoupled from the calling activity. That basically allows more code reusability and for permission related checks to live in a utility class that handles it all. Most Activities or controls only care about knowing if a permission was granted or not so that they can react accordingly. eg. with this approach all the widgets will be able to simply plug in into the `PermissionsUtil` and receive a callback without cluttering the `FormEntryActivity`. 

#### Are there any risks to merging this code? If so, what are they?
There are no risks associated with merging this code because the target version isn't set to 23. So the default behaviour of the permissions system will take place.

#### Do we need any specific form for testing your changes? If so, please attach one.
No. 
#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Not necessarily. Permissions are pretty self explanatory and users are accustomed to them popping up in other apps. 

Reviews welcome :)

jd-alexander committed:
added initial permissions implementation for storage.

jd-alexander committed:
added initial permissions implementation for storage.

jd-alexander committed:
moved view instantiation code above permission check since onResume is making calls to some elements and grabbing views from the layout won't be impacted by result of permission check.

jd-alexander committed:
moved view instantiation code above permission check since onResume is making calls to some elements and grabbing views from the layout won't be impacted by result of permission check.

jd-alexander committed:
moved view instantiation code above permission check since onResume is making calls to some elements and grabbing views from the layout won't be impacted by result of permission check.

jd-alexander committed:
moved view instantiation code above permission check since onResume is making calls to some elements and grabbing views from the layout won't be impacted by result of permission check.

jd-alexander committed:
added permission check to ActivityLogger so that it doesn't start running until the storage permissions are granted.

jd-alexander committed:
added permission check to ActivityLogger so that it doesn't start running until the storage permissions are granted.

jd-alexander committed:
Added GrantPermissionRule so that the tests on API 23 and above would be granted all the necessary permissions before they run.

jd-alexander committed:
Added GrantPermissionRule so that the tests on API 23 and above would be granted all the necessary permissions before they run.

jd-alexander committed:
removed unnecessary null pointer checks.

jd-alexander committed:
removed unnecessary null pointer checks.

jd-alexander committed:
added NonNull annotations to trigger null warnings; replaces null checks.

jd-alexander committed:
added NonNull annotations to trigger null warnings; replaces null checks.

jd-alexander committed:
did a revert on ActivityLogger since it's current implementation won't get triggered until dirs are created. Applied permission rules and checks to all Content Providers and Activities that can be launched from an intent.

jd-alexander committed:
did a revert on ActivityLogger since it's current implementation won't get triggered until dirs are created. Applied permission rules and checks to all Content Providers and Activities that can be launched from an intent.

jd-alexander committed:
code fixes and cleanups.

jd-alexander committed:
code fixes and cleanups.

jd-alexander commented:
hmmm I am not seeing some of these checkstyle issues when I run it locally. Gonna figure out what's causing that.

jd-alexander committed:
fixed checkstyle issues.

jd-alexander committed:
fixed checkstyle issues.

jd-alexander committed:
added logging to the content providers so they tell when permissions aren't available.

jd-alexander committed:
added logging to the content providers so they tell when permissions aren't available.

jd-alexander commented:
Lint is failing the build due to tons of depreciated results :cry: 
I guess because I opened these files they were marked as modified.

jd-alexander committed:
removed due to no other declarants of these permissions being present. they were added as a fix to permission dialogs that weren't showing up during debugging.

jd-alexander committed:
removed due to no other declarants of these permissions being present. they were added as a fix to permission dialogs that weren't showing up during debugging.

jd-alexander committed:
temporarily disabled lint to ensure all other checks are passing. Lint is failing due to depreciated code in several other classes.

jd-alexander committed:
temporarily disabled lint to ensure all other checks are passing. Lint is failing due to depreciated code in several other classes.

lognaturel commented:
Awesome! 

The lint errors are all related to missing permissions and the deprecation warnings you're seeing aren't breaking the build. I think the way to go will be to only change the targetSdk for verification and then change it back before merging until all the permission checks are eventually added across a series of PRs. I believe that should work, right?

jd-alexander commented:
Oh yeah! I just looked at the output properly and saw that after the warnings several other tasks are executed! I am gonna make the sdk change now. Btw, how did you spot the permissions error? I am looking at the log but I haven't seen it as yet.

On a side note, I have been having some issues with checkstyle. When I first started i tried to clone Collect on my mac but LeakCanary wasn't being found. Some repo issue so I cloned on Windows and now I have like 60K line ending errors :sweat_smile: so I need to solve that to get more value out of checkstyle locally.

jd-alexander committed:
refactored initialization statements that were pulled from onCreate of FormEntryActivity

jd-alexander committed:
refactored initialization statements that were pulled from onCreate of FormEntryActivity

lognaturel commented:
> how did you spot the permissions error?

There are two types of output from the quality tools -- console output that's warnings and status and the actual tool report. At the end you get a path to the report when you run locally. You can access them in CI by going to Artifacts and then clicking on the report you want. You may need a CircleCI account, I'm not sure!

> Some repo issue so I cloned on Windows and now I have like 60K line ending errors 😅 so I need to solve that to get more value out of checkstyle locally.

Those pesky Windows line endings! If you identify a good, clean way to fix those, it would be a good thing to add to the README. You're not the first one to get that problem.

huangyz0918 commented:
> When I first started i tried to clone Collect on my mac but LeakCanary wasn't being found.

@jd-alexander I think trying to `rebuild project ` or `clean project`  can fix this error. If this approach is useless, you can comment the LeakCanary temporarily in order to run your app in Mac .

jd-alexander commented:
thanks for the info @lognaturel I am gonna investigate a fix for it. 

@huangyz0918 Yes, that's an approach I could take for sure.

jd-alexander committed:
performed rebase and resolved conflicts.

jd-alexander committed:
peformed merge

lognaturel commented:
@jd-alexander When you have a chance, could you please fix the merge conflicts, rebase on master and force push to get a clean diff? You should be able to do something like
`git fetch upstream`
`git rebase upstream/master`
`git push --force`

At that point, only commits related to this PR should show up in the commit list and in the diff.

Might be good to also make sure the PR description reflects the latest state of things. For example, the targetSdk isn't changed to 23 anymore and I think we agree that keeping all the permissions checks in a long-running branch will result in too much effort to fix merge conflicts. Merging one at a time should be relatively safe.

Thank you!

jd-alexander commented:
Will do that. I was using the rebase with Github fork functionality in Android Studio and I was wondering why I wasn't getting the same result. Will use this method instead. Thanks for the info.

jd-alexander committed:
Merge branch 'master' into storage_runtime_permissions

jd-alexander committed:
Update sd.xml

Changed color to that of `?colorControlNormal` to adhere to proper theme support.

jd-alexander commented:
Hi @lognaturel so I was doing the rebase on my machine with the commands above and had to manually walk through about 29 conflicts; took me a few hours and even at the end I had some fixes to make. Then I realized that the resolve conflicts feature on here only picked up 4 conflicts and once I resolved those and corrected a minor error I was good to go. Do you know how Github does that? Would like to know if there's some form of command or strategy that would allow me to accomplish the same via git bash.

lognaturel commented:
> took me a few hours and even at the end I had some fixes to make

Oh no! Please post in Slack if something unexpected like that happens so we can discuss. When you do a rebase you're replaying all of your commits on top of the target branch so each applied commit is an opportunity for a conflict. What probably happened is that the conflicts were spread across most of your commits. On the other hand, doing a merge effectively applies all changes at once. I guess in this case it would have been easier to merge in master first and then rebase on it so the old commits weren't part of the diff.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -196,7 +198,7 @@
     private static final int SAVING_IMAGE_DIALOG = 3;
 
-    private boolean autoSaved;
+    private boolean autoSaved = false;
```
This shouldn't make a difference -- booleans default to false, right?

dcbriccetti replied:
Yes, the ` = false` is noise. Noise must be eliminated. :-)

jd-alexander replied:
Eliminating noise 😄

jd-alexander commented:
Oh I understand. Yes, I should have posted there! I just caught up trying to complete until enough time went by for me to realize this isn't the best approach. Lesson learnt though. Since the conflicts have been resolved this can be merged in after testing?

lognaturel commented:
For QA, the `targetSdkVersion` will need to be changed to 23: https://github.com/jd-alexander/collect/blob/1db6152160f53abb745699435573c7b334c90010/collect_app/build.gradle#L74. There should be no change for devices running API < 23. For devices running API > 23, on first launch after a clean install, the app should ask for SD card write permissions. If granted, the app behavior should be unchanged. If not granted, the app should fail to launch.

The trickiness comes in when a user [revokes permissions](https://www.addictivetips.com/android/how-to-revoke-individual-permissions-for-an-app-in-android-6-0/) while running Collect. In those cases we need to be sure that Collect prompts for permissions instead of just crashing.

jd-alexander committed:
removed redundant boolean initialization

mmarciniak90 commented:
@jd-alexander I noticed crash in case:
1. User uses a device with Android 6.0
2. User opens Collect and sees permission dialog
3. User clicks on `DENY` and then OK on dialog with information
4. User opens Collect again and clicks on `ALLOW`
5. Collect crashes - User sees dialog 'Unfortunately, ODK Collect has stopped.`

I have 2 devices with Android 6.0 and crash is on both devices.
When I open Collect once more everything is ok.

Revoking permissions does not cause a crash - verified on Android 6.0, 7.0, 8.1

jd-alexander commented:
@mmarciniak90 I was able to reproduce the issue on an Android 6.0 device. Working on a fix.

jd-alexander committed:
fixed a bug that occurred due to requestWindowFeature causing a crash

RequestWindowFeature should always be called   before making any view related calls such as SetContentView.

jd-alexander committed:
moved super call of onCreate to start of the method

jd-alexander committed:
Merge branch 'master' into storage_runtime_permissions

jd-alexander commented:
@mmarciniak90 Issue was resolved! Good to go now.

mmarciniak90 commented:
@jd-alexander previous crash was fixed but I noticed case with revoking permission when the app crashes. 
1. The user opens the app and allows permission
2. The user can fill a form or open some view like eg. `Edit Saved Form` or `Get Blank Form`
3. Permission is unselected and the user returns to Collect
4. If the user goes to `Get Blank Form` view, app crashes - dialog 'Unfortunately, ODK Collect has stopped.` is visible
5. If the user goes to `Delete Saved Form` view, the app also crashes but dialog 'Unfortunately, ODK Collect has stopped.` is not visible

jd-alexander commented:
Okay I am gonna take a look at these.

jd-alexander committed:
Added storage permission checks to all activities so when the permission is revoked it's handled gracefully.

Entry point activities were able to handle storage permissions but all other activities weren't so they crashed when the user manually revoked the permission. Now when a user revokes a permission the app shows a dialog about it and then quits. This approach is taken because the revoking of storage permissions is an edge case in a sense and if an user actually does it, it's easier for them to restart the app via an entry point than to have permission granted code in all activities.

jd-alexander committed:
Merge remote-tracking branch 'origin/storage_runtime_permissions' into storage_runtime_permissions

# Conflicts:
#	collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

jd-alexander committed:
Removed unused import.

jd-alexander committed:
Merge branch 'master' into storage_runtime_permissions

jd-alexander commented:
Resolved the issues! What was happening is that when the permissions were revoked the activities without permission checks didn't know how to handle the lack of the storage permission so they just crashed. What happens now is when permissions are revoked and you resume your activity a dialog shows and tells you that you have re-enable them when you are ready by restarting the app. Adding permission checks for every screen seemed a bit too much since all the activities that are used to open the app actually have them and if a user decides to revoke such a critical permission then they 1) shouldn't have any problem re-opening the app whe they are ready to grant them back 2) won't revoke it again since they want to use the app and it's basically mandatory.

mmarciniak90 commented:
*Tested with success!*

Verified on Android 8.1, 7.0, 6.0 and also on checked on Android 5.1, 4.4, 4.2, 4.1

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

lognaturel committed:
Merge branch 'master' into storage_runtime_permissions

lognaturel committed:
Use string notation to match new convention

lognaturel commented:
Hurray! One step closer to not being booted from the Play Store in November. 😄
